xbmc/filesystem/SpecialProtocol.cpp

```diff
@@ -28,12 +28,402 @@
+#if defined (TARGET_WINDOWS)
+  path = URIUtils::AddFileToFolder(CWIN32Util::GetProfilePath(), "userdata");
+#endif
+#if defined(TARGET_POSIX) && !defined(TARGET_DARWIN)
+  if (g_application.PlatformDirectoriesEnabled())
+  {
+    path = GetHomePath() + "/userdata";
+  }
+  else
+  {
+    path = URIUtils::AddFileToFolder(GetXBMCPath(), "portable_data/userdata");
+  }
+#endif
+
+#ifdef TARGET_DARWIN
+  if (g_application.PlatformDirectoriesEnabled())
+  {
+    // REVIEW: Make a get function for userHome calc
+    std::string userHome;
+    if (getenv("HOME"))
+      userHome = getenv("HOME");
+    else
+      userHome = "/root";
+
+#if defined(TARGET_DARWIN_IOS)
+    path = userHome + "/" + CDarwinUtils::GetAppRootFolder() + "/" + CCompileInfo::GetAppName() + "/userdata";
+#else
+    path = userHome + "/Library/Application Support/" + CCompileInfo::GetAppName() + "/userdata";
+#endif
+  }
+  else
+  {
+    path = URIUtils::AddFileToFolder(GetXBMCPath(), "portable_data");
+  }
+
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetMasterProfilePath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetMasterProfilePath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetXBMCPath()
+{
+  std::string path;
+#if defined (TARGET_WINDOWS)
+  path = CUtil::GetHomePath();
+#endif
+
+#if defined(TARGET_POSIX) && !defined(TARGET_DARWIN)
+  path = INSTALL_PATH;
+  /* Check if binaries and arch independent data files are being kept in
+  * separate locations. */
+  if (!CDirectory::Exists(URIUtils::AddFileToFolder(path, "userdata")))
+  {
+    /* Attempt to locate arch independent data files. */
+    auto appBinPath = CUtil::GetHomePath("KODI_BIN_HOME");
+    path = CUtil::GetHomePath(appBinPath);
+    if (!g_application.PlatformDirectoriesEnabled())
+    {
+      URIUtils::AddSlashAtEnd(path);
+    }
+  }
 #endif
+#if defined (TARGET_DARWIN)
+  path = CUtil::GetHomePath();
+  if (!g_application.PlatformDirectoriesEnabled())
+  {
+    URIUtils::AddSlashAtEnd(path);
+  }
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetXBMCPath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetXBMCPath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetXBMCBinPath()
+{
+  std::string path;
+#ifdef TARGET_WINDOWS
+  path = GetXBMCPath();
+#endif
+
+#if defined(TARGET_POSIX) && !defined(TARGET_DARWIN)
+  path = CUtil::GetHomePath("KODI_BIN_HOME");
+  if (!g_application.PlatformDirectoriesEnabled())
+  {
+    URIUtils::AddSlashAtEnd(path);
+  }
+#endif
+#ifdef TARGET_DARWIN
+  path = GetXBMCPath();
+  if (!g_application.PlatformDirectoriesEnabled())
+  {
+    URIUtils::AddSlashAtEnd(path);
+  }
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetXBMCBinPath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetXBMCBinPath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetXBMCBinAddonPath()
+{
+  std::string path;
+#ifdef TARGET_WINDOWS
+  path = GetXBMCPath("/addons");
```

Comment:
you don't need a slash here `URIUtils::AddFileToFolder` will make sure a slash is added.